### PR TITLE
Fallback to ArrayPool in Kestrel

### PR DESCRIPTION
**Pull request by BrennanConroy**
Shows improvements when writing large Json payloads.
e.g.
Scenario: Modified https://github.com/aspnet/Benchmarks/blob/a10414a273ec5aa5d1482134499f366093c560ba/src/Benchmarks/Middleware/JsonMiddleware.cs#L37 with a 400k json string
RPS: 4.3k -> 10.5k with Kestrel ArrayPool addition
Allocation rate: 5 Billion B/sec -> 568 Million B/sec

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs [hunk 1]
```diff
@@ -740,6 +740,12 @@ private void AddSegment(int sizeHint = 0)
             _currentSegment = owner.Memory;
             _currentSegmentOwner = owner;
         }
+        else if (sizeHint <= MemoryPool<byte>.Shared.MaxBufferSize)
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs:743]: What are the values of `_memoryPool.MaxBufferSize` and `MemoryPool<byte>.Shared.MaxBufferSize`? How much of a gap is there between these values?

Comment (BrennanConroy) [reply src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs:743]: 4k vs 2GB

**Review by Tratcher — APPROVED** [review]

**Comment by danmoseley** [comment]
After this change how much of the remaining allocations are of buffers (direct or inside the pool)? Hopefully it's minimal?

**Comment by BrennanConroy** [comment]
> After this change how much of the remaining allocations are of buffers (direct or inside the pool)? Hopefully it's minimal?

Could you clarify what you mean?

File: src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs (hunk 1 above)
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs:743]: nit: Reduce duplication by adding a static method and call it twice. e.g.

```
static bool TryRentBuffer(MemoryPool<byte> pool, int sizeHint, ref Memory<byte> buffer, ref IMemoryOwner<byte>? owner)
{
    if (sizeHint <= pool.MaxBufferSize)
    {
        owner = pool.Rent(sizeHint);
        buffer = owner.Memory;
        return true;
    }

    return false;
}
```

Then call with `_memoryPool` and `MemoryPool<byte>.Shared`.

File: src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs [hunk 2]
```diff
@@ -740,6 +740,12 @@ private void AddSegment(int sizeHint = 0)
...
+        {
```
Comment (adityamandaleeka) [inline src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs:744]: May want to add a comment about why we do this.

**Review by adityamandaleeka — APPROVED** [review]
Nice!

**Review by davidfowl — APPROVED** [review]

**Comment by danmoseley** [comment]
> Could you clarify what you mean

We now use the pool, so we expect (and get) fewer allocations of these buffers. There's still plenty of allocations in the scenario it sounds like, I'm curious whether much of that is due to not finding a pooled buffer here and having to allocate one, vs unrelated allocations.
(Only if you happen to know)

**Comment by BrennanConroy** [comment]
Ah. I don't have a list of the exact allocations, but if you (royal you) look at the allocation rate chart (pass `--chart` when running the benchmark) you can see that almost all the allocations occur at the beginning of the benchmark run. So that implies that most allocations are being done via the pool now. Since this sort of scenario (large writes) is something I'm interested in looking into, I'm sure there are other improvements to be made in future changes. (And we've already played with some experimental changes that improve perf on top of this change)

**Comment by danmoseley** [comment]
Gotcha - thanks

**Pull request merged**